File: src/py/serve.py
"""Worker script for serving model predictions for non-training purposes."""
import argparse
import asyncio
import json
from pathlib import Path
from typing import NamedTuple, Optional, TypedDict, Union, cast

import tensorflow as tf
import zmq
import zmq.asyncio

from .models.dqn_model import DQNModel
from .models.drqn_model import DRQNModel
from .models.utils.greedy import decode_action_rankings
from .models.utils.q_value import decode_q_values
from .utils.state import decode_state


class Ready(TypedDict):
    """Sent by parent process to ensure the worker has started."""

    type: str
    """Must be `"ready"`."""


class Ack(TypedDict):
    """Sent to parent process to acknowledge Ready."""

    type: str
    """Must be `"ack"`."""


class ModelRequest(TypedDict):
    """
    Model prediction request protocol. Encoded state data is sent in a separate
    buffer.
    """

    type: str
    """Must be `"model"`."""

    id: int
    """Differentiates multiple requests from the same client."""

    key: str
    """
    Key for looking up hidden states or other context used to continue a battle.
    """


class ModelReply(TypedDict):
    """Model prediction reply protocol."""

    type: str
    """Must be `"model"`."""

    id: int
    """Differentiates multiple requests from the same client."""

    key: str
    """Same key that was passed from the ModelRequest."""

    ranked_actions: list[str]
    """Action names ranked by the model."""

    q_values: Optional[dict[str, float]]
    """Maps action names to predicted Q-values. Optional debug info."""


class CleanupRequest(TypedDict):
    """
    Request protocol for cleaning up stored prediction context from a battle.
    """

    type: str
    """Must be `"cleanup"`."""

    key: str
    """Key identifier for the stored context."""


class Pending(NamedTuple):
    """Represents a pending request for the server."""

    routing_id: bytes
    """Id of the socket to send the result back to."""

    req: ModelRequest
    """Request json."""

    state: tf.Tensor
    """Encoded tensor state input."""


async def server(
    model_path: Path, sock_id: str, max_batch: int, debug_outputs=False
):
    """Serve worker function."""

    model: Union[DQNModel, DRQNModel] = tf.keras.models.load_model(
        model_path,
        custom_objects={"DQNModel": DQNModel, "DRQNModel": DRQNModel},
        compile=False,
    )
    is_recurrent = model.__class__.__name__ == "DRQNModel"
    print(f"Loaded model from {model_path}")

    ctx = zmq.asyncio.Context.instance()
    sock = ctx.socket(zmq.ROUTER)
    sock.setsockopt(zmq.LINGER, 0)
    sock.setsockopt(zmq.ROUTER_MANDATORY, 1)
    sock.setsockopt(zmq.SNDHWM, 0)
    sock.bind(f"ipc:///tmp/psai-serve-socket-{sock_id}")

    poller = zmq.asyncio.Poller()
    poller.register(sock, zmq.POLLIN)

    # Initialize owning worker.
    routing_id, ready_bytes = await sock.recv_multipart()
    ready: Ready = json.loads(ready_bytes)
    assert ready["type"] == "ready"
    ack_msg: Ack = {"type": "ack"}
    await sock.send_multipart([routing_id, json.dumps(ack_msg).encode()])

    if is_recurrent:
        hiddens: dict[str, list[tf.Tensor]] = {}

    # Naive serve loop. Only meant for lightweight demos.
    print("Serving model")
    while True:
        # Gather pending model requests.
        pending: dict[str, Pending] = {}
        while (
            max_batch <= 0 or len(pending) < max_batch
        ) and await poller.poll(timeout=0 if len(pending) > 0 else None):
            msg = await sock.recv_multipart(flags=zmq.DONTWAIT, copy=False)
            routing_id = msg[0].bytes
            req: Union[ModelRequest, CleanupRequest] = json.loads(msg[1].bytes)
            key = req["key"]
            if req["type"] == "cleanup":
                if is_recurrent:
                    del hiddens[key]
                continue
            req = cast(ModelRequest, req)
            assert req["type"] == "model"
            assert key not in pending

            state = tf.convert_to_tensor(
                decode_state(msg[2].buffer), dtype=tf.float32
            )
            pending[key] = Pending(routing_id=routing_id, req=req, state=state)
            if is_recurrent and key not in hiddens:
                hiddens[key] = DRQNModel.new_hidden()

        # Execute batch.
        keys = [*pending.keys()]
        batch_state = tf.stack([pending[key].state for key in keys])
        if not is_recurrent:
            if debug_outputs:
                ranked_actions, q_values = model.greedy_with_q(batch_state)
            else:
                ranked_actions = model.greedy(batch_state)
                q_values = None
        else:
            # Need to also handle sequence dim and hidden states.
            batch_state = tf.expand_dims(batch_state, axis=1)
            batch_hidden = list(
                map(tf.stack, zip(*(hiddens[key] for key in keys)))
            )
            if debug_outputs:
                ranked_actions, new_hiddens, q_values = model.greedy_with_q(
                    batch_state, batch_hidden
                )
                q_values = tf.squeeze(q_values, axis=1)
            else:
                ranked_actions, new_hiddens = model.greedy(
                    batch_state, batch_hidden
                )
                q_values = None
            ranked_actions = tf.squeeze(ranked_actions, axis=1)
            hiddens |= zip(keys, map(list, zip(*map(tf.unstack, new_hiddens))))

        ranked_actions = decode_action_rankings(ranked_actions)
        if q_values is not None:
            q_values = decode_q_values(q_values)
        else:
            q_values = [None] * len(pending)

        # Send replies.
        for key, action, q_dict in zip(keys, ranked_actions, q_values):
            routing_id, req, _ = pending[key]
            rep: ModelReply = {
                "type": "model",
                "id": req["id"],
                "key": key,
                "ranked_actions": action,
                "q_values": q_dict,
            }
            await sock.send_multipart([routing_id, json.dumps(rep).encode()])


def main():
    """Main function."""
    parser = argparse.ArgumentParser()
    parser.add_argument("model_path", type=Path)
    parser.add_argument("sock_id", type=str)
    parser.add_argument("--max-batch", default=32, type=int)
    parser.add_argument("--debug-outputs", action="store_true")

    args = parser.parse_args()

    asyncio.run(
        server(
            model_path=args.model_path,
            sock_id=args.sock_id,
            max_batch=args.max_batch,
            debug_outputs=args.debug_outputs,
        )
    )


if __name__ == "__main__":
    main()

File: src/ts/model/serve.ts
import {ChildProcess, spawn} from "child_process";
import * as path from "path";
import * as readline from "readline";
import * as zmq from "zeromq";
import {Action} from "../battle/agent";
import {ReadonlyBattleState} from "../battle/state";
import {stateEncoder} from "../battle/state/encoder";
import {UsageStats} from "../battle/usage";
import {Logger} from "../utils/logging/Logger";

// eslint-disable-next-line @typescript-eslint/naming-convention
const PYTHONPATH = path.resolve(__dirname, "..", "..", "..");

/** Ensures that the model server has been started. */
interface Ready {
    type: "ready";
}

/** Sent by the model server to acknowledge {@link Ready}. */
interface Ack {
    type: "ack";
}

/**
 * Model prediction request protocol. Encoded state data is sent in a separate
 * buffer.
 */
interface ModelRequest {
    type: "model";
    /** Differentiates multiple requests from the same client. */
    id: number;
    /**
     * Key for looking up hidden states or other context used to continue a
     * battle.
     */
    key: string;
}

/** Model prediction reply protocol. */
interface ModelReply {
    type: "model";
    /** Differentiates multiple requests from the same client. */
    id: number;
    /** All possible actions ranked by the model. */
    // eslint-disable-next-line @typescript-eslint/naming-convention
    ranked_actions: Action[];
    /** Predicted Q-values for each action. */
    // eslint-disable-next-line @typescript-eslint/naming-convention
    q_values?: {[T in Action]: number};
}

/** Request protocol for cleaning up stored prediction context from a battle. */
interface CleanupRequest {
    type: "cleanup";
    /** Key identifier for the stored context. */
    key: string;
}

/** Contains prediction data from the model server. */
export interface ModelPrediction {
    /** All possible actions ranked by the model. */
    rankedActions: Action[];
    /** Predicted Q-values for each action. Optional debug info. */
    qValues?: {[T in Action]: number};
}

/** Server class for handling model predictions. */
export class ModelServer {
    /** Manages the Python subprocess. */
    private serveProcess?: ChildProcess;
    /** Used for communicating with the subprocess. */
    private socket?: zmq.Dealer;

    /** Callbacks for pending requests. */
    private readonly pendingRequests = new Map<
        number,
        (rep: ModelReply) => void
    >();
    /** Counter for numbering requests. */
    private requestCount = 0;

    /** Task for extracting async agent replies. */
    private pullerPromise?: Promise<void>;

    /** Whether the server is closed. */
    private closed = false;

    /**
     * Creates a ModelServer.
     *
     * @param modelPath Path to the model to serve.
     * @param socketId Id to use for the socket.
     * @param maxBatch Max batch size for inference.
     * @param logger Logger object.
     */
    public constructor(
        public readonly modelPath: string,
        public readonly socketId: string,
        public readonly maxBatch: number,
        private readonly logger: Logger,
    ) {}

    /** Starts the server. */
    public async start(): Promise<void> {
        this.serveProcess = spawn(
            "python",
            [
                "-u", // Unbuffered stdout/stderr.
                "-m",
                "src.py.serve",
                this.modelPath,
                this.socketId,
                "--max-batch",
                this.maxBatch.toString(),
                "--debug-outputs",
            ],
            // eslint-disable-next-line @typescript-eslint/naming-convention
            {env: {...process.env, PYTHONPATH}, stdio: "pipe"},
        );

        const rl = readline.createInterface({
            input: this.serveProcess.stdout!,
            terminal: false,
        });
        rl.on("line", line => this.logger.debug(line));

        const rlErr = readline.createInterface({
            input: this.serveProcess.stderr!,
            terminal: false,
        });
        rlErr.on("line", line => this.logger.error(line));

        const context = new zmq.Context();
        this.socket = new zmq.Dealer({context, linger: 0});
        this.socket.connect(`ipc:///tmp/psai-serve-socket-${this.socketId}`);

        // Ensure worker has started successfully.
        const ready: Ready = {type: "ready"};
        await this.socket.send(JSON.stringify(ready));

        this.socket.receiveTimeout = 60_000 /*60s*/;
        const [ackBuf] = await this.socket.receive();
        this.socket.receiveTimeout = -1;

        const ack = JSON.parse(ackBuf.toString()) as Ack;
        if (ack.type !== "ack") {
            throw new Error(`Invalid ack received: ${ackBuf.toString()}`);
        }

        this.pullerPromise = this.puller();
        this.pullerPromise.catch(e =>
            this.logger.error(
                `Uncaught error in puller: ${(e as Error).stack ?? e}`,
            ),
        );
    }

    /** Closes the server. */
    public close() {
        this.serveProcess?.kill();
        this.socket?.close();
        this.closed = true;
    }

    /**
     * Sends a model prediction request and awaits its response.
     *
     * @param key Key for looking up stored context used to continue a battle.
     * Should be unique for each battle perspective.
     * @param state Battle state representation.
     * @param usage Usage stats to impute for state encoder.
     * @param smoothing Confidence smoothing used in imputation algorithm.
     * @returns Prediction data from the model server.
     */
    public async predict(
        key: string,
        state: ReadonlyBattleState,
        usage?: UsageStats,
        smoothing?: number,
    ): Promise<ModelPrediction> {
        const req: ModelRequest = {type: "model", id: ++this.requestCount, key};
        const stateData = new Float32Array(stateEncoder.size);
        stateEncoder.encode(stateData, {state, usage, smoothing});

        const replyPromise = new Promise<ModelReply>(res =>
            this.pendingRequests.set(req.id, res),
        ).finally(() => this.pendingRequests.delete(req.id));

        await this.ensureSocket().send([JSON.stringify(req), stateData.buffer]);
        const reply = await replyPromise;
        return {
            rankedActions: reply.ranked_actions,
            ...(reply.q_values && {qValues: reply.q_values}),
        };
    }

    /**
     * Cleans up stored context from battle predictions.
     *
     * @param key Context key.
     */
    public async cleanup(key: string): Promise<void> {
        const req: CleanupRequest = {type: "cleanup", key};
        await this.ensureSocket().send([JSON.stringify(req)]);
    }

    /** Extracts async predictions from the subprocess. */
    private async puller(): Promise<void> {
        while (!this.closed) {
            const [replyBuffer] = await this.ensureSocket().receive();
            const reply = JSON.parse(replyBuffer.toString()) as ModelReply;
            if (reply.type !== "model") {
                throw new Error(`Unknown reply type '${reply.type}'`);
            }
            const cb = this.pendingRequests.get(reply.id);
            if (!cb) {
                throw new Error(`Unexpected reply id ${reply.id}`);
            }
            cb(reply);
        }
    }

    /** Asserts initialization precondition. */
    private ensureSocket(): zmq.Dealer {
        if (!this.socket) {
            throw new Error("ModelServer wasn't started or was closed");
        }
        return this.socket;
    }
}
